### ISSUE TITLE
[Epic] Autoscaling clusters

### ISSUE THREAD

**Issue by benesch** [MaterializeInc/materialize#22568]
## Product outcome

Clusters scale automatically in response to changes in workload.

Today, users must manually choose the initial scale for their cluster (via `CREATE CLUSTER ... SIZE`) and size it up and down in response to load (via `ALTER CLUSTER ... SET (SIZE ...)`).

## Discovery

The bulk of the discovery work is still TODO.

## Work items

TODO.

## See also

  * https://github.com/MaterializeInc/materialize/issues/13870

## Action log

* **22 October 2023.** Issue created.

**Comment by edude03** [comment]
Fwiw, the Kubernetes Vertical autoscaler would implement the reactive out of the box if we let it. I'm not sure though it'd be well integrated into our product however

**Comment by jubrad** [comment]
I'm thinking about autoscaling in a post [graceful reconfig](https://github.com/MaterializeInc/materialize/issues/20010) world. It seems once the `WAIT UNTIL CAUGHT UP` feature lands two things become possible. 
1.  We can use the `CREATE CLUSTER foo REPLICATION FACTOR 3, MIN SIZE 'xsmall', MAX SIZE 'xlarge'` to create a schedule that looks at the cluster every so many seconds and gracefully resizes it.
2. A customer could write their own app which subscribe to custom views they write which look at mz_introspection data and spits out a recommended size then issues a ddl gracefully updating the cluster on changes. It would probably want to have some anti-flapping.

**Comment by benesch** [comment]
Agreed on all counts! (1) is exactly how I envision building autoscaling.
